### PR TITLE
Native npm

### PR DESCRIPTION
**Pull request by Postyy**

### CODE REVIEW

File: artifactory/cli.go [hunk 1]
```diff
@@ -587,6 +592,17 @@ func GetCommands() []cli.Command {
 				return pipDepsTreeCmd(c)
 			},
 		},
+		{
+			Name:         "npm-config",
+			Flags:        getGlobalConfigFlag(),
```
Comment (yahavi) [inline artifactory/cli.go]:
```suggestion
			Flags:        getGlobalConfigFlag(),
                      Aliases:      []string{"npmc"},
```

File: artifactory/utils/prompt/promptutils.go [hunk 2]
```diff
@@ -174,3 +179,59 @@ func GetRepositories(resolveRes *viper.Viper, repoTypes ...utils.RepoType) ([]st
 
 	return utils.GetRepositories(artAuth, repoTypes...)
 }
+
+func CreateBuildConfig(global bool, confType utils.ProjectType) error {
```
Comment (yahavi) [inline artifactory/utils/prompt/promptutils.go]: I suggest creating npmconfig.go file and put this function there, as do other techs - go, pip, gradle and maven.

File: artifactory/cli.go [hunk 3]
```diff
@@ -2872,10 +3033,27 @@ func overrideIntIfSet(field *int, c *cli.Context, fieldName string) {
 	}
 }
 
-func validateBuildParams(buildConfig *utils.BuildConfiguration) error {
-	if (buildConfig.BuildName == "" && buildConfig.BuildNumber != "") || (buildConfig.BuildName != "" && buildConfig.BuildNumber == "") || (buildConfig.Module != "" && buildConfig.BuildName == "" && buildConfig.BuildNumber == "") {
-		return errors.New("The build-name, build-number and module options cannot be sent separately.")
+func validateCommonContext(c *cli.Context) error {
```
Comment (yahavi) [inline artifactory/cli.go]: I couldn't see any usage for that function

File: artifactory/cli.go [hunk 4]
```diff
@@ -1587,14 +1604,56 @@ func nugetDepsTreeCmd(c *cli.Context) error {
 }
 
 func npmInstallCmd(c *cli.Context) error {
+	configFilePath, err := isProjectConfExist(utils.Npm)
```
[the remark this answers is not in the document]
Comment (Postyy) [reply artifactory/cli.go]: Change:
utils.GetProjectConfFilePath -> utils.IsProjectConfFilePath 
isProjectConf -> getProjectConf

[the remark this answers is not in the document]
Comment (Postyy) [reply artifactory/cli.go]: 'isProjectConfExist' was replaced by 'utils.GetProjectConfFilePath'

File: artifactory/cli.go [hunk 5]
```diff
@@ -1603,20 +1662,49 @@ func npmInstallCmd(c *cli.Context) error {
...
+	npmInstallArgs, err := shellwords.Parse(c.String("npm-args"))
+	if err != nil {
+		return err
+	}
+	npmCmd.SetThreads(threads).SetBuildConfiguration(buildConfiguration).SetRepo(c.Args().Get(0)).SetNpmArgs(npmInstallArgs).SetRtDetails(rtDetails)
 
 	return commands.Exec(npmCmd)
 }
 
 func npmCiCmd(c *cli.Context) error {
```
[the remark this answers is not in the document]
Comment (Postyy) [reply artifactory/cli.go]: Yes, merged both of them to single a func.

File: artifactory/cli.go [hunk 6]
```diff
@@ -1587,14 +1604,56 @@ func nugetDepsTreeCmd(c *cli.Context) error {
...
+		if c.NArg() < 1 {
+			return cliutils.PrintHelpAndReturnError("Wrong number of arguments.", c)
+		}
+		args, err := shellwords.Parse(strings.Join(extractCommand(c), " "))
+		if err != nil {
+			return errorutils.CheckError(err)
+		}
+		// Validates the npm command. If a config file is found, the only flags that can be used are threads, build-name, build-number and module.
+		// Otherwise, throw an error.
+		if err := validateCommand(args, getNpmBasicFlages()); err != nil {
```
[the remark this answers is not in the document]
Comment (Postyy) [reply artifactory/cli.go]: No, we would like the user to decide the number of threads.

File: artifactory/cli.go [hunk 7]
```diff
@@ -1603,20 +1662,49 @@ func npmInstallCmd(c *cli.Context) error {
...
+
+	if configFilePath != "" {
+		// Found a config file. Continue as native command.
+		args, err := shellwords.Parse(strings.Join(extractCommand(c), " "))
+		if err != nil {
+			return errorutils.CheckError(err)
+		}
+		// Validates the npm command. If a config file is found, the only flags that can be used are threads, build-name, build-number and module.
+		// Otherwise, throw an error.
+		if err := validateCommand(args, getNpmBasicFlages()); err != nil {
```
[the remark this answers is not in the document]
Comment (Postyy) [reply artifactory/cli.go]: https://github.com/jfrog/jfrog-cli/pull/498#discussion_r349913769

File: artifactory/cli.go [hunk 8]
```diff
@@ -854,15 +870,19 @@ func getDockerFlags() []cli.Flag {
...
 			Name:  "npm-args",
 			Usage: "[Optional] A list of npm arguments and options in the form of \"--arg1=value1 --arg2=value2\"` `",
 		},
 	}
 	npmFlags = append(npmFlags, getBaseFlags()...)
-	npmFlags = append(npmFlags, getServerIdFlag())
+	return append(npmFlags, getServerIdFlag())
+}
+
+func getNpmCommonFlags() []cli.Flag {
```
[the remark this answers is not in the document]
Comment (Postyy) [reply artifactory/cli.go]: Done.

File: artifactory/cli.go [hunk 9]
```diff
@@ -1630,6 +1718,51 @@ func npmCiCmd(c *cli.Context) error {
...
+	if configFilePath != "" {
+		// Found a config file. Continue as native command.
+		if c.NArg() < 1 {
+			return cliutils.PrintHelpAndReturnError("Wrong number of arguments.", c)
+		}
+		args, err := shellwords.Parse(strings.Join(extractCommand(c), " "))
+		if err != nil {
+			return errorutils.CheckError(err)
+		}
+		// Validates the npm command. If a config file is found, the only flags that can be used are threads, build-name, build-number and module.
```
[the remark this answers is not in the document]
Comment (Postyy) [reply artifactory/cli.go]: https://github.com/jfrog/jfrog-cli/pull/498#discussion_r349913769

File: artifactory/cli.go [hunk 10]
```diff
@@ -1630,6 +1718,51 @@ func npmCiCmd(c *cli.Context) error {
...
+		// Otherwise, throw an error.
+		if err := validateCommand(args, getNpmBasicFlages()); err != nil {
+			return err
+		}
+		npmCmd := npm.NewNpmPublishCommand()
+		log.Info("Running npm-publish Install.")
+		// Read config file.
```
[the remark this answers is not in the document]
Comment (Postyy) [reply artifactory/cli.go]: Done.

File: artifactory/cli.go [hunk 11]
```diff
@@ -1642,7 +1775,11 @@ func npmPublishCmd(c *cli.Context) error {
 	if err != nil {
 		return err
 	}
-	npmPublicCmd.SetBuildConfiguration(buildConfiguration).SetRepo(c.Args().Get(0)).SetNpmArgs(c.String("npm-args")).SetRtDetails(rtDetails)
+	npmPublicArgs, err := shellwords.Parse(c.String("npm-args"))
```
[the remark this answers is not in the document]
Comment (Postyy) [reply artifactory/cli.go]: Those args related to ``npm pack``

File: artifactory/cli.go [hunk 12]
```diff
@@ -1691,6 +1828,21 @@ func shouldSkipGoFlagParsing() bool {
 	return exists
 }
 
+func shouldSkipNpmFlagParsing() bool {
+	// This function is executed by code-congsta, regardless of the CLI command being executed.
+	// There's no need to run the code of this function, if the command is not "jfrog rt npm-install || npm".
```
[the remark this answers is not in the document]
Comment (Postyy) [reply artifactory/cli.go]: Done.

File: artifactory/cli.go [hunk 13]
```diff
@@ -1691,6 +1828,21 @@ func shouldSkipGoFlagParsing() bool {
...
+	if len(os.Args) < 3 || (os.Args[2] != "npm-install" && os.Args[2] != "npmi" && os.Args[2] != "npm-ci" &&
+		os.Args[2] != "npmci" && os.Args[2] != "npm-publish" && os.Args[2] != "npmp") {
```
[the remark this answers is not in the document]
Comment (Postyy) [reply artifactory/cli.go]: Got a better idea:
regexp.MustCompile(`npmi | npmci |npm-install...`)

File: artifactory/cli.go [hunk 14]
```diff
@@ -1751,8 +1903,9 @@ func goNativeCmd(c *cli.Context, configFilePath string) error {
 		return cliutils.PrintHelpAndReturnError("Wrong number of arguments.", c)
 	}
 	args := extractCommand(c)
-	// Validate the command
-	if err := validateGoNativeCommand(args); err != nil {
+	// Validates the go command. If a config file is found, the only flags that can be used are build-name, build-number and module.
```
[the remark this answers is not in the document]
Comment (Postyy) [reply artifactory/cli.go]: No, those args are related for legacy use

File: artifactory/cli.go (hunk 1 above)
Comment (Postyy) [reply artifactory/cli.go]: Done.

File: artifactory/utils/prompt/promptutils.go (hunk 2 above)
Comment (Postyy) [reply artifactory/utils/prompt/promptutils.go]: Done.

File: artifactory/cli.go (hunk 3 above)
Comment (Postyy) [reply artifactory/cli.go]: Playground tests, removed.

File: artifactory/commands/npm/publish.go [hunk 15]
```diff
@@ -35,15 +35,58 @@ type NpmPublishCommand struct {
...
+		if err != nil {
+			return err
+		}
+		// Extract resolution params.
+		deployerParams, err := utils.GetRepoConfigByPrefix(npc.configFilePath, utils.ProjectConfigDeployerPrefix, vConfig)
+		if err != nil {
+			return err
+		}
+		_, filteredNpmArgs, buildConfiguration, err := utils.ExtractNpmOptionsFromArgs(npc.NpmPublishCommandArgs.npmArgs)
+		RtDetails, err := deployerParams.RtDetails()
```
Comment (eyalbe4) [inline artifactory/commands/npm/publish.go]: *RtDetails* should be probably renamed to *rtDetails*.

File: artifactory/commands/npm/install.go [hunk 16]
```diff
@@ -52,20 +52,58 @@ type NpmCommandArgs struct {
...
+
+func (nic *NpmInstallCommand) Run() error {
+	log.Info("Running npm Install.")
+	// Read config file.
+	log.Debug("Preparing to read the config file", nic.configFilePath)
+	vConfig, err := utils.ReadConfigFile(nic.configFilePath, utils.YAML)
+	if err != nil {
+		return err
+	}
+	// Extract resolution params.
```
Comment (eyalbe4) [inline artifactory/commands/npm/install.go]: Is this comment correct?

File: artifactory/cli.go [hunk 17]
```diff
@@ -587,6 +594,18 @@ func GetCommands() []cli.Command {
 				return pipDepsTreeCmd(c)
 			},
 		},
+		{
+			Name:         "npm-config",
```
Comment (eyalbe4) [inline artifactory/cli.go]: I suggest placing the **npm-config** command as the first npm command (similarly to the **pip0config** command).
Let's dp the same for the **go-config** command.

File: artifactory/cli.go [hunk 18]
```diff
@@ -854,15 +873,20 @@ func getDockerFlags() []cli.Flag {
 	return flags
 }
 
-func getNpmCommonFlags() []cli.Flag {
+//These flags are not valid for native npm command
```
Comment (eyalbe4) [inline artifactory/cli.go]: //These flags... --> // This flag...
...native npm command --> ...native npm commands.

File: artifactory/cli.go [hunk 19]
```diff
@@ -854,15 +873,20 @@ func getDockerFlags() []cli.Flag {
...
+func getNpmLegacyFlages() []cli.Flag {
 	npmFlags := []cli.Flag{
 		cli.StringFlag{
 			Name:  "npm-args",
 			Usage: "[Optional] A list of npm arguments and options in the form of \"--arg1=value1 --arg2=value2\"` `",
```
Comment (eyalbe4) [inline artifactory/cli.go]: I suggest adding **[Deprecated]** to all deprecated options (both npm and go) as follows:
Usage: "[Optional] [Deprecated] A list of npm arguments and options in the form of \"--arg1=value1 --arg2=value2\"` `",

File: artifactory/cli.go [hunk 20]
```diff
@@ -1691,6 +1771,20 @@ func shouldSkipGoFlagParsing() bool {
 	return exists
 }
 
+func shouldSkipNpmFlagParsing() bool {
+	// This function is executed by code-congsta, regardless of the CLI command being executed.
```
Comment (eyalbe4) [inline artifactory/cli.go]: code-congsta --> code-gangsta

File: main_test.go [hunk 21]
```diff
@@ -134,3 +138,43 @@ func initArtifactoryCli() {
 		configArtifactoryCli = createConfigJfrogCLI(cred)
 	}
 }
+
+func testCreateConfFile(dirs []string, resolver, deployer string, t *testing.T, confType artifactoryUtils.ProjectType) error {
```
Comment (eyalbe4) [inline main_test.go]: The function name probably does not match its functionality.
I'd also suggest adding a description for this function.

Comment (Or-Gabay) [reply main_test.go]: testCreateConfFile  -> createConfigFileForTest

**Pull request merged**